### ISSUE TITLE
A bug in BatchNorm Op ?

### ISSUE THREAD

**Issue by jimxinbo** [dmlc/mxnet#1053]
Hi there, great work coding up the batchnorm layer. I can see it follows mostly to the BN paper except for using moving-mean, var for efficience, that is OK!

However, I think there a bug in the backward code. 

In line 187-191 of batch_norm-inl.h, after computing gmean correctly, the code screwed it up by gmean += tmp. Not sure why to compute the tmp at all, becasue the following assignment to grad_in is correct without this extra tmp. 

Seems the effect of this extra tmp will not be huge because it was devided by m 2 time, but mathematically it is wrong according to both my math and the BN paper. 

I maybe wrong and I hope I was, let me know.

**Comment by piiswrong** [comment]
I didn't write the code but you can verify by either 1) check the gradient with gradient checker in test/python/unittests/check_utils.py or 2) compare it with CuDNNBatchNorm
@antinucleon

**Comment by jimxinbo** [comment]
Thanks piiswrong, always in time. Sorry I could not test it mysefl now. But honestly, I think it is pretty obvious, just ask related people to check about it. Anyway, It's clearly more of carelessness than a math mistake. 

Note that, in the Inception-BN example, the corresponding m in BN is quit large and the difference can be rather minor as to affect the results. So maybe no one noticed it.

**Comment by winstywang** [comment]
Looking into it.

**Comment by winstywang** [comment]
@jimxinbo Check P4 of http://arxiv.org/pdf/1502.03167.pdf. The gradient w.r.t. \mu has two terms, the tmp corresponds to the second term.

**Comment by jimxinbo** [comment]
@winstywang thx Naiyan, interesting. 

Check out this icml paper http://jmlr.csail.mit.edu/proceedings/papers/v37/ioffe15.pdf

I will look into it later but you can also check out why the difference.

**Comment by jimxinbo** [comment]
Hmm at a first glance though, seems the arxiv version took into account mu_b in sigma_b.  Maybe the arxiv one is correct.
